src/machine.ts

```diff
@@ -0,0 +1,66 @@
+import { Path, PathType } from './path';
+
+export class State {
+  x: number;
+  y: number;
+  z: number;
+  e: number;
+  tool: number;
+  units: 'mm' | 'in';
+
+  static get initial(): State {
+    const state = new State();
+    Object.assign(state, { x: 0, y: 0, z: 0, e: 0, tool: 0, units: 'mm' });
+    return state;
+  }
+}
+
+export class Machine {
+  paths: Path[];
+  state: State;
+
+  constructor(state?: State) {
+    this.paths = [];
+    this.state = state || State.initial;
+  }
+
+  isPlanar(): boolean {
+    return (
+      this.paths.find(
+        (path) =>
+          path.travelType === PathType.Extrusion && path.vertices.some((_, i, arr) => i % 3 === 2 && arr[i] !== arr[2])
+      ) === undefined
+    );
+  }
```

Comment:
Also look kinda expensive. Should we do this check in the parser?

Reply:
If all the parser cares about is the syntax of gcode, this logic does not belong in the parser. However, if the gcode has a special comment about layers, that's another story. 

The role of the interpreter is to encapsulate anything that has to do with understanding what is actually going on. 

I had another implementation that had some tradeoff I did not want to make. It was directly in the G1 method. It had the advantage of not looping a 2nd time through all commands. But also had the disadvantage to add a condition no matter if it was already determined to be non-planar. 

I did not implement memoization because of processing gcode in chunks, but could be a short term solution.

Reply:
As discussed in discord, optimisations will come in a follow-up PR